File: 234. Palindrome Linked List/isPalindrome.py
from typing import Optional, List
# Definition for singly-linked list.
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next

# helper function to generate LinkedList from array
def arrayToLinkedList(arr: Optional[List]) -> Optional[ListNode]:
  if not arr:
    return None

  dummyHead = ListNode(0)
  curr = dummyHead

  for i in range(len(arr)):
    node = ListNode(arr[i])
    curr.next = node
    curr = curr.next
  
  return dummyHead.next

class Solution:
  def isPalindrome(self, head: Optional[ListNode]) -> bool:
    slow = head
    fast = head

    while fast and fast.next:
      fast = fast.next.next
      slow = slow.next

    prev = None
    while slow:
      tempNext = slow.next
      slow.next = prev
      prev = slow
      slow = tempNext

    left, right = head, prev
    while right:
      if left.val != right.val:
        return False
      
      left = left.next
      right = right.next

    return True
  
head = arrayToLinkedList([1, 2, 2, 1])
solution = Solution()
print('isPalindrome: ', solution.isPalindrome(head))

File: 234. Palindrome Linked List/isPalindrome.js
// Definition for singly-linked list.
class ListNode {
  constructor(val = 0, next = null) {
    this.val = val;
    this.next = next;
  }
}

// helper function to generate LinkedList from array
const arrayToLinkedList = (arr) => {
  if (!arr || !arr.length) {
    return null;
  }

  const dummyHead = new ListNode(0);
  let curr = dummyHead;

  for (let i = 0; i < arr.length; i++) {
    let node = new ListNode(arr[i]);
    curr.next = node;
    curr = curr.next;
  }

  return dummyHead.next;
}

/**
 * @param {ListNode} head
 * @return {boolean}
 */
const isPalindrome = (head) => {
  let slow = head;
  let fast = head;

  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
  }

  let prev = null;
  while (slow) {
    let temp = slow.next;
    slow.next = prev;
    prev = slow
    slow = temp;
  }

  let left = head;
  let right = prev;
  while (right) {
    if (left.val !== right.val) {
      return false;
    }
    left = left.next;
    right = right.next;
  }

  return true;
};

const head = arrayToLinkedList([1, 2, 2, 1]);
console.log('isPalindrome: ', isPalindrome(head));
